crates/uv-resolver/src/resolution/display.rs

```diff
@@ -336,42 +334,72 @@ fn combine_extras<'dist>(graph: &IntermediatePetGraph<'dist>) -> RequirementsTxt
+                let index = next.add_node(dist.clone());
+                entry.insert(index);
+            }
         }
     }
 
-    // Verify that the package markers are more general than the extra markers.
-    if cfg!(debug_assertions) {
-        for index in graph.node_indices() {
-            let DisplayResolutionGraphNode::Dist(dist) = &graph[index] else {
-                continue;
-            };
-            let combined_markers = next[inverse[&dist.version_id()]].markers.clone();
-            let mut package_markers = combined_markers.clone();
-            package_markers.or(graph[index].markers().clone());
-            assert_eq!(
-                package_markers,
-                combined_markers,
-                "{} {:?} {:?}",
-                dist.version_id(),
-                dist.extras,
-                dist.markers.try_to_string()
-            );
+    // Re-add the edges to the reduced graph.
+    for edge in graph.edge_indices() {
+        let (source, target) = graph.edge_endpoints(edge).unwrap();
+        let DisplayResolutionGraphNode::Dist(source_node) = &graph[source] else {
+            continue;
+        };
+        let DisplayResolutionGraphNode::Dist(target_node) = &graph[target] else {
+            continue;
+        };
+        let source = inverse[&version_marker(source_node)];
+        let target = inverse[&version_marker(target_node)];
+
+        next.update_edge(source, target, ());
+    }
+
+    next
+}
+
+/// Reduce the graph, such that all nodes for a single package are combined, with extras
+/// removed.
+///
+/// For example, `flask`, `flask[async]`, and `flask[dotenv]` should be reduced into a single
+/// `flask` node, with a conjunction of their markers.
+///
+/// We also remove the root node, to simplify the graph structure.
+fn strip_extras<'dist>(graph: &IntermediatePetGraph<'dist>) -> RequirementsTxtGraph<'dist> {
+    let mut next = RequirementsTxtGraph::with_capacity(graph.node_count(), graph.edge_count());
+    let mut inverse = FxHashMap::with_capacity_and_hasher(graph.node_count(), FxBuildHasher);
+
+    // Re-add the nodes to the reduced graph.
+    for index in graph.node_indices() {
+        let DisplayResolutionGraphNode::Dist(dist) = &graph[index] else {
+            continue;
+        };
+
+
+        // In the `requirements.txt` output, we want a flat installation list, so we need to use
+        // the reachability markers instead of the edge markers.
+        match inverse.entry(dist.version_id()) {
+            std::collections::hash_map::Entry::Occupied(entry) => {
+                let index = *entry.get();
+                let node: &mut RequirementsTxtDist = &mut next[index];
+                node.extras.clear();
+            }
+            std::collections::hash_map::Entry::Vacant(entry) => {
+                let index = next.add_node(dist.clone());
+                entry.insert(index);
+            }
         }
     }
```

Comment:
I could make this a trait to DRY up the above but... I don't know, it seems like a lot of complexity.